motif/match.go

```diff
@@ -0,0 +1,304 @@
+	for row = 0; row < len(m); row++ {
+		answer = answer + "[\t"
+		for column = 0; column < len(m[row]); column++ {
+			currBaseString = dna.BaseToString(m[row][column].Base)
+			answer = answer + fmt.Sprintf("%f:%s\t", m[row][column].Value, currBaseString)
+		}
+		answer = answer + "]\n"
+	}
+	answer = answer + "\n"
+
+	return answer
+}
+
+// buildKmerHash produces a hash mapping 2bit encoded kmer sequences to their corresponding motif score for a PositionMatrix.
+// Only kmers with a motif score above an input thresholdProportion are stored in the
+func buildKmerHash(p PositionMatrix, thresholdProportion float64) map[uint64]float64 {
+	var answer = make(map[uint64]float64)
+	var currSeq = ConsensusSequence(p, false)
+	consensusValue, _, couldScoreConsensus := ScoreWindow(p, currSeq.Seq, 0)
+	if !couldScoreConsensus {
+		log.Fatalf("Error in buildKmerHash. Could not score consensus sequence.")
+	}
+	var threshold float64 = thresholdProportion * consensusValue
+	var rankMatrix [][]rankTensorElement = initializeRankTensor(p)
+	var currRankVector = make([]int, len(p.Mat[0])) //intialize to all zeros, representing the consensus sequence.
+	var consensusKey uint64 = dnaTwoBit.BasesToUint64RightAln(currSeq.Seq, 0, len(currSeq.Seq))
+	answer[consensusKey] = consensusValue
+
+	for column := 0; column < len(rankMatrix[0]); column++ {
+		//make child sequence and rank vector
+		currSeq.Seq[column] = rankMatrix[1][column].Base
+		currRankVector[column] = 1
+		//call recursive part
+		recursiveCheckKmers(answer, currSeq.Seq, rankMatrix, consensusValue, currRankVector, column, threshold)
+		//revert child sequence to parent before next recursive call
+		currSeq.Seq[column] = rankMatrix[0][column].Base
+		currRankVector[column] = 0
+	}
+
+	return answer
+}
+
+// recursiveCheckKmers is a helper function of buildKmerHash, which searches for all kmer permutations of a PostionMatrix consensus sequence
+// and adds all kmers with a motif score above an input threshold to a map. This implementation uses dynamic programming to search all kmers efficiently.
+func recursiveCheckKmers(answer map[uint64]float64, currSeq []dna.Base, rankMatrix [][]rankTensorElement, parentValue float64, rankVector []int, index int, threshold float64) {
+	//first score sequence
+	//we look at parent score, and we look at changed base.
+	currValue := parentValue + rankMatrix[rankVector[index]][index].Value - rankMatrix[rankVector[index]-1][index].Value
+	// if we pass threshold. Everything else is in this if, so if we fail threshold, we just return
+	if currValue >= threshold {
+		//if we've passed, we add current sequence to the map
+		var currKey uint64 = dnaTwoBit.BasesToUint64RightAln(currSeq, 0, len(currSeq))
+		answer[currKey] = currValue
+		for i := index; i < len(rankMatrix[0]); i++ { //generate children to the right of the current index.
+			if rankVector[i] < 3 { //we only have to generate children if the rank vector is less than 3 at this position (only 4 base identities)
```

Comment:
Would it be a small efficiency boost if the if and for lines are swapped so the for loop only gets executed if the "if" is met?

Reply:
This would change the logic. The 'if' will be true for some loop iterations and false for others